gitpod-io/gitpod
# Support more regional workspace cluster

## Description
Today workspace cluster choice is guided solely by the location of the application cluster the workspace is started from. This PR introduces an admission preference (introduced in #6164) which adds more regional support.

The current implementation in this branch is a rather rough first start. It does two things:
1. introduce the `region` admission preferences on workspace clusters. Each such admission preference carries a "round trip time endpoint" which is used to measure the RTT of a user to a particular region. We assume that if there are multiple clusters within the same "region" that their "RTT endpoints" yield equivalent results.
2. make the `start workspace` page measure the RTT and update the cluster preference for the user. This part I'm least happy with, because it's
  - super spammy and makes many more requests than it should
  - comes too late: only after the first workspace is running will this be done

It's particularly with point 2 that I'd need some support @gitpod-io/engineering-meta :)

## Related Issue(s)

Fixes #5596

## How to test
1. remove the default cluster registration by editing the `WSMAN_CFG` env var on server setting it to `W10K` (base64 for `[]`)
2. register a the ws-manager of the preview environment with `gpctl clusters register ...` make sure it's governed
3. register a ws-manager of another preview environment
4. use `gpctl clusters update admission-preference add ...` to point one cluster to a "slow" RTT endpoint, e.g. `ws-us25.gitpod.io`, and one to a "fast" RTT endpoint, e.g. `ws-eu25.gitpod.io` (depending on where in the world you're located)
5. start a workspace and observe the correct cluster choice

## Release Notes

```release-note
Support more regional workspace cluster
```

## Documentation
